state/migrations/20240517104423_device_list_table.go

```diff
@@ -0,0 +1,188 @@
+	var userID string
+	var deviceID string
+	var data []byte
+	// every N seconds log an update
+	updateFrequency := time.Second * 2
+	lastUpdate := time.Now()
+	i := 0
+	for {
+		// logging
+		i++
+		if time.Since(lastUpdate) > updateFrequency {
+			logger.Info().Msgf("%d/%d process device list data", i, count)
+			lastUpdate = time.Now()
+		}
+
+		if err := tx.QueryRow(
+			`FETCH NEXT FROM device_data_migration_cursor`,
+		).Scan(&userID, &deviceID, &data); err != nil {
+			if err == sql.ErrNoRows {
+				// End of rows.
+				break
+			}
+			return err
+		}
+
+		//  * deserialise the CBOR
+		result, err := deserialiseCBOR(data)
+		if err != nil {
+			return err
+		}
+
+		//  * transfer the device lists to the new device lists table
+		var deviceListRows []state.DeviceListRow
+		for targetUser, targetState := range result.DeviceLists.New {
+			deviceListRows = append(deviceListRows, state.DeviceListRow{
+				UserID:       userID,
+				DeviceID:     deviceID,
+				TargetUserID: targetUser,
+				TargetState:  targetState,
+				Bucket:       state.BucketNew,
+			})
+		}
+		for targetUser, targetState := range result.DeviceLists.Sent {
+			deviceListRows = append(deviceListRows, state.DeviceListRow{
+				UserID:       userID,
+				DeviceID:     deviceID,
+				TargetUserID: targetUser,
+				TargetState:  targetState,
+				Bucket:       state.BucketSent,
+			})
+		}
+		if len(deviceListRows) == 0 {
+			continue
+		}
+		chunks := sqlutil.Chunkify(5, state.MaxPostgresParameters, state.DeviceListChunker(deviceListRows))
+		for _, chunk := range chunks {
+			var placeholders []string
+			var vals []interface{}
+			listChunk := chunk.(state.DeviceListChunker)
+			for i, deviceListRow := range listChunk {
+				placeholders = append(placeholders, fmt.Sprintf("($%d,$%d,$%d,$%d,$%d)",
+					i*5+1,
+					i*5+2,
+					i*5+3,
+					i*5+4,
+					i*5+5,
+				))
+				vals = append(vals, deviceListRow.UserID, deviceListRow.DeviceID, deviceListRow.TargetUserID, deviceListRow.TargetState, deviceListRow.Bucket)
+			}
+			query := fmt.Sprintf(
+				`INSERT INTO syncv3_device_list_updates(user_id, device_id, target_user_id, target_state, bucket) VALUES %s`,
+				strings.Join(placeholders, ","),
+			)
+			_, err = tx.ExecContext(ctx, query, vals...)
+			if err != nil {
+				return fmt.Errorf("failed to bulk insert: %s", err)
+			}
+		}
+
+		//  * delete the device lists from the CBOR and update
+		result.DeviceLists = OldDeviceLists{
+			New:  make(internal.MapStringInt),
+			Sent: make(internal.MapStringInt),
+		}
+		data, err := cbor.Marshal(result)
+		if err != nil {
+			return err
+		}
+		_, err = tx.ExecContext(ctx, `UPDATE syncv3_device_data SET data=$1 WHERE user_id=$2 AND device_id=$3`, data, userID, deviceID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func downDeviceListTable(ctx context.Context, tx *sql.Tx) error {
+	// no-op: we'll drop the device list updates but still work correctly as new/sent are still in the cbor but are empty
```

Comment:
Shouldn't the down migration still be done for completeness?
Or is the theory here that future syncs will rebuild the device list updates in the cbor well enough that we don't really care to do the down migration?

Also - does this need to drop the `syncv3_device_list_updates` table? Or does that happen elsewhere?